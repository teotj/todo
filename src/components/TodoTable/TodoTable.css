.root {
  width: 100%;
  margin-bottom: 10px;
  margin-left: 10px;
  border-collapse: collapse;
}

.theader {
  .TodosCell {
    width: 85%;
  }

  .ActionsCell {
    width: 15%;
  }
}

.tbody {
}

.tr {
}

.titleCompleted {
  text-decoration: line-through;
}

.titleActive {
}

.action {
}

.complete {
}

.delete {
}

.error {
  color: #c23030;
}
