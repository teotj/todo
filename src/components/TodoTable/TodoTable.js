import React from 'react';
import PropTypes from 'prop-types';
import { Button, Callout, HTMLTable, Intent, Tooltip } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';

import css from './TodoTable.css';

const TodoTable = ({ items, onComplete, onDelete }) => {
  if (items.length === 0) {
    return (
      <Callout intent={Intent.DANGER} className={css.error}>
        Todo list is either empty or filter returns no results.
      </Callout>
    );
  }

  const columns = ['Todos', 'Actions'];
  const completeButton = (id, isCompleted) => (
    <Tooltip content="Mark complete" disabled={isCompleted}>
      <Button
        minimal
        disabled={isCompleted}
        icon={IconNames.TICK_CIRCLE}
        intent={Intent.SUCCESS}
        className={css.complete}
        onClick={(e) => onComplete(e, id)}
      />
    </Tooltip>
  );
  const deleteButton = (id) => (
    <Tooltip content="Delete todo">
      <Button
        minimal
        icon={IconNames.BAN_CIRCLE}
        intent={Intent.DANGER}
        className={css.delete}
        onClick={(e) => onDelete(e, id)}
      />
    </Tooltip>
  );

  return (
    <HTMLTable className={css.root} interactive>
      <thead className={css.theader}>
        <tr>
          {columns.map((column) => (
            <th key={column} className={css[`${column}Cell`]}>{column}</th>
          ))}
        </tr>
      </thead>
      <tbody className={css.tbody}>
        {items.map(({ title, isCompleted, id }) => {
          const titleCn = isCompleted ? css.titleCompleted : css.titleActive;
          return (
            <tr key={id} className={css.tr}>
              <td className={titleCn}>{title}</td>
              <td className={css.action}>
                {completeButton(id, isCompleted)}
                {deleteButton(id)}
              </td>
            </tr>
          );
        })}
      </tbody>
    </HTMLTable>
  );
};

TodoTable.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string,
    isCompleted: PropTypes.bool,
    id: PropTypes.string
  })).isRequired,
  onComplete: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired
};

export default TodoTable;
