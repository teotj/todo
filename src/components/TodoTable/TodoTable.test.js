import React from 'react';
import { shallow } from 'enzyme';

import TodoTable from './TodoTable';

const renderTodoTable = (props = {}) => {
  const defaultItems = [
    { title: 'TodoA', id: '-id1-', isCompleted: true },
    { title: 'TodoB', id: '-id2-', isCompleted: false }
  ];
  const defaultOnComplete = () => {};
  const defaultOnDelete = () => {};
  return (
    <TodoTable
      items={defaultItems}
      onComplete={defaultOnComplete}
      onDelete={defaultOnDelete}
      {...props}
    />
  );
};

const shallowRenderTodoTable = (props) => shallow(renderTodoTable(props));

describe('<TodoTable>', () => {
  it('Should render', () => {
    const actual = shallowRenderTodoTable();
    expect(actual).toMatchSnapshot();
  });

  it('Should render empty results', () => {
    const items = [];
    const actual = shallowRenderTodoTable({ items });
    expect(actual).toMatchSnapshot();
  });

  it('Should call onComplete', () => {
    const onComplete = jest.fn();
    const actual = shallowRenderTodoTable({ onComplete });
    actual.find('.complete').at(0).simulate('click'); // As there are 2 items, call the first one
    expect(onComplete).toBeCalledWith(undefined, '-id1-'); // Undefined for the 'e' event
  });

  it('Should call onDelete', () => {
    const onDelete = jest.fn();
    const actual = shallowRenderTodoTable({ onDelete });
    actual.find('.delete').at(1).simulate('click');
    expect(onDelete).toBeCalledWith(undefined, '-id2-');
  });
});
