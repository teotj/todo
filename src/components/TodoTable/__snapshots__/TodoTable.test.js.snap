// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoTable> Should render 1`] = `
<HTMLTable
  className="root"
  interactive={true}
>
  <thead
    className="theader"
  >
    <tr>
      <th
        className="TodosCell"
        key="Todos"
      >
        Todos
      </th>
      <th
        className="ActionsCell"
        key="Actions"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="tbody"
  >
    <tr
      className="tr"
      key="-id1-"
    >
      <td
        className="titleCompleted"
      >
        TodoA
      </td>
      <td
        className="action"
      >
        <Blueprint3.Tooltip
          content="Mark complete"
          disabled={true}
          hoverCloseDelay={0}
          hoverOpenDelay={100}
          transitionDuration={100}
        >
          <Blueprint3.Button
            className="complete"
            disabled={true}
            icon="tick-circle"
            intent="success"
            minimal={true}
            onClick={[Function]}
          />
        </Blueprint3.Tooltip>
        <Blueprint3.Tooltip
          content="Delete todo"
          hoverCloseDelay={0}
          hoverOpenDelay={100}
          transitionDuration={100}
        >
          <Blueprint3.Button
            className="delete"
            icon="ban-circle"
            intent="danger"
            minimal={true}
            onClick={[Function]}
          />
        </Blueprint3.Tooltip>
      </td>
    </tr>
    <tr
      className="tr"
      key="-id2-"
    >
      <td
        className="titleActive"
      >
        TodoB
      </td>
      <td
        className="action"
      >
        <Blueprint3.Tooltip
          content="Mark complete"
          disabled={false}
          hoverCloseDelay={0}
          hoverOpenDelay={100}
          transitionDuration={100}
        >
          <Blueprint3.Button
            className="complete"
            disabled={false}
            icon="tick-circle"
            intent="success"
            minimal={true}
            onClick={[Function]}
          />
        </Blueprint3.Tooltip>
        <Blueprint3.Tooltip
          content="Delete todo"
          hoverCloseDelay={0}
          hoverOpenDelay={100}
          transitionDuration={100}
        >
          <Blueprint3.Button
            className="delete"
            icon="ban-circle"
            intent="danger"
            minimal={true}
            onClick={[Function]}
          />
        </Blueprint3.Tooltip>
      </td>
    </tr>
  </tbody>
</HTMLTable>
`;

exports[`<TodoTable> Should render empty results 1`] = `
<Blueprint3.Callout
  className="error"
  intent="danger"
>
  Todo list is either empty or filter returns no results.
</Blueprint3.Callout>
`;
